src/index.ts

```diff
@@ -0,0 +1,775 @@
+/**
+ * A Result model, loosely based on the Rust result type.
+ *
+ * @module
+ */
+
+/**
+ * `Result<T, E>` is used for returning and propagating errors.
+ *
+ * It can be constructed with {@link Ok} or {@link Err}.
+ *
+ * @typeParam T - The type of the `Ok` value.
+ * @typeParam E - The type of the `Err` value.
+ */
+export class Result<T, E> {
+  /** @internal */
+  // biome-ignore lint/complexity/noUselessConstructor: hiding from documentation
+  constructor() {}
+
+  /**
+   * Returns `true` if the result is `Ok`.
+   *
+   * @example
+   *
+   * ```typescript
+   * const x: Result<number, string> = Ok(2)
+   * x.isOk() === true
+   *
+   * const y: Result<number, string> = Err('hey')
+   * y.isOk() === false
+   * ```
+   */
+  isOk(): boolean {
+    throw new Error('Abstract Result')
+  }
+
+  /**
+   * Returns `true` if the result is `Ok` and the contained value matches a
+   * predicate.
+   *
+   * @param pred - The predicate to test the contained value
+   * @typeParam T - The type of the contained value
+   *
+   * @example
+   *
+   * ```typescript
+   * const x: Result<number, string> = Ok(2)
+   * x.isOkAnd((v) => v > 1) === true
+   *
+   * const y: Result<number, string> = Ok(0)
+   * y.isOkAnd((v) => v > 1) === false
+   *
+   * const z: Result<number, string> = Err('hey')
+   * z.isOkAnd((v) => v > 1) === false
+   * ```
+   */
+  isOkAnd(pred: (val: T) => boolean): boolean {
+    throw new Error('Abstract Result')
+  }
+
+  /**
+   * Returns `true` if the result is `Err`.
+   *
+   * @example
+   *
+   * ```typescript
+   * const x: Result<number, string> = Ok(2)
+   * x.isErr() === false
+   *
+   * const y: Result<number, string> = Err('hey')
+   * y.isErr() === true
+   * ```
+   */
+  isErr(): boolean {
+    throw new Error('Abstract Result')
+  }
+
+  /**
+   * Returns `true` if the result is `Err` and the contained error matches
+   * a predicate.
+   *
+   * @param pred - The predicate to test the contained error
+   * @typeParam E - The type of the contained error
+   *
+   * @example
+   *
+   * ```typescript
+   * const x: Result<number, string> = Ok(2)
+   * x.isErrAnd((e) => e === 'hey') === false
+   *
+   * const y: Result<number, string> = Err('howdy')
+   * y.isErrAnd((e) => e === 'hey') === false
+   *
+   * const z: Result<number, string> = Err('hey')
+   * z.isErrAnd((e) => e === 'hey') === true
+   * ```
+   */
+  isErrAnd(pred: (val: E) => boolean): boolean {
+    throw new Error('Abstract Result')
+  }
+
+  /**
+   * Maps a `Result<T, E>` to `Result<U, E>` by applying a function to
+   * the contained `Ok` value, leaving an `Err` value untouched.
+   *
+   * @param fn - A function that maps `T` to `U`
+   * @typeParam T - The type of the contained value
+   * @typeParam U - The type of the new contained value
```

Comment:
I’m assuming the use of the next character for the new type is convention.

Reply:
It's what I have seen in a few places. It could also be `T -> T2`, but those feel much messier to me.